concordium-consensus/src/Concordium/GlobalState/Persistent/TreeState.hs

```diff
@@ -902,33 +920,30 @@ instance
             let nonce = transactionNonce tr
                 sender = accountAddressEmbed (transactionSender tr)
             anft <- use (skovPersistentData . transactionTable . ttNonFinalizedTransactions . at' sender . non emptyANFT)
-            if anft ^. anftNextNonce == nonce
+            let nfn = anft ^. anftMap . at' nonce . non Map.empty
+                wmdtr = WithMetadata{wmdData = tr, ..}
+            if Map.member wmdtr nfn
                 then do
-                    let nfn = anft ^. anftMap . at' nonce . non Map.empty
-                        wmdtr = WithMetadata{wmdData = tr, ..}
-                    if Map.member wmdtr nfn
-                        then do
-                            -- Remove any other transactions with this nonce from the transaction table.
-                            -- They can never be part of any other block after this point.
-                            forM_ (Map.keys (Map.delete wmdtr nfn)) $
-                                \deadTransaction -> skovPersistentData . transactionTable . ttHashMap . at' (getHash deadTransaction) .= Nothing
-                            -- Mark the status of the transaction as finalized, and remove the data from the in-memory table.
-                            ss <- deleteAndFinalizeStatus wmdHash
-                            -- Update the non-finalized transactions for the sender
-                            skovPersistentData
-                                . transactionTable
-                                . ttNonFinalizedTransactions
-                                . at' sender
-                                ?= ( anft
-                                        & (anftMap . at' nonce .~ Nothing)
-                                        & (anftNextNonce .~ nonce + 1)
-                                   )
-                            return ss
-                        else do
-                            logErrorAndThrowTS $ "Tried to finalize transaction which is not known to be in the set of non-finalized transactions for the sender " ++ show sender
+                    -- Remove any other transactions with this nonce from the transaction table.
+                    -- They can never be part of any other block after this point.
+                    forM_ (Map.keys (Map.delete wmdtr nfn)) $
+                        \deadTransaction -> skovPersistentData . transactionTable . ttHashMap . at' (getHash deadTransaction) .= Nothing
+                    -- Mark the status of the transaction as finalized, and remove the data from the in-memory table.
+                    ss <- deleteAndFinalizeStatus wmdHash
+                    -- Remove the transaction from the non finalized transactions.
+                    -- If there are no non-finalized transactions left then remove the entry
+                    -- for the sender in @ttNonFinalizedTransactions@.
```

Comment:
**Nitpick:** Introduce a function for removing non-finalized transactions from the transaction table.

The code below seems to deal with the internal structure of the transaction table and only behaves correct, since the `_anftNextNonce` is no longer part of the type. I think it would be better to factor this out and keep it close to the type for better 'separation of concerns'.

Reply:
good idea